File: scripts/cpp/test_script2.cpp
#include <iostream>
#include <cmath>
#include <chrono>
#include <thread>
#include <array>
#include <Eigen/Dense>
#include <Eigen/Geometry>

#include <trossen_vr/vr_manager.hpp>
#include <trossen_vr/vr_types.hpp>

#include "libtrossen_arm/trossen_arm.hpp"


Eigen::Matrix4d vec6_to_T(const std::array<double, 6>& v6) {
    Eigen::Vector3d p(v6[0], v6[1], v6[2]);
    Eigen::Vector3d rvec(v6[3], v6[4], v6[5]);
    double angle = rvec.norm();
    Eigen::Matrix3d R = Eigen::Matrix3d::Identity();
    if (angle > 1e-8) {
        Eigen::Vector3d axis = rvec / angle;
        R = Eigen::AngleAxisd(angle, axis).toRotationMatrix();
    }
    Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
    T.block<3,3>(0,0) = R;
    T.block<3,1>(0,3) = p;
    return T;
}

std::array<double, 6> T_to_vec6(const Eigen::Matrix4d& T) {
    std::array<double, 6> v6;
    Eigen::Vector3d p = T.block<3,1>(0,3);
    v6[0] = p.x(); v6[1] = p.y(); v6[2] = p.z();

    Eigen::Matrix3d R = T.block<3,3>(0,0);
    Eigen::AngleAxisd aa(R);
    Eigen::Vector3d rvec = aa.axis() * aa.angle();
    v6[3] = rvec.x(); v6[4] = rvec.y(); v6[5] = rvec.z();
    return v6;
}

std::optional<std::array<double,6>> pose_to_vec6(const std::optional<trossen_vr::VRPose>& pose) {
    if (!pose.has_value()) return std::nullopt;
    std::array<double, 6> v;
    v[0] = pose->position[0];
    v[1] = pose->position[1];
    v[2] = pose->position[2];
    v[3] = pose->rotation[0];
    v[4] = pose->rotation[1];
    v[5] = pose->rotation[2];
    return v;
}



// -------------------------------------
// MAIN — MANUAL MODE
// -------------------------------------
int main() {
    using namespace trossen_vr;

    std::vector<double> START_POSE = {0, M_PI/12, M_PI/12, 0, 0, 0};
    std::vector<double> IDLE_POSE  = {0, 0, 0, 0, 0, 0};

    // ---------------------------
    // Robot setup
    // ---------------------------
    trossen_arm::TrossenArmDriver driver;
    driver.configure(trossen_arm::Model::wxai_v0,
                     trossen_arm::StandardEndEffector::wxai_v0_follower,
                     "192.168.1.2", true);

    driver.set_arm_modes(trossen_arm::Mode::position);
    driver.set_gripper_mode(trossen_arm::Mode::position);
    driver.set_arm_positions(START_POSE, 3.0, true);

    std::cout << "Robot ready.\n";



    // ---------------------------
    // VRManager (manual mode)
    // ---------------------------
    VRManager::Config cfg;
    cfg.server_port = 5432;
    VRManager vr_manager(cfg);

    vr_manager.start();

    std::cout << "VR Manager started in MANUAL mode.\n";



    // ---------------------------
    // State for teleop alignment
    // ---------------------------
    bool pause_teleop = false;
    std::optional<std::array<double,6>> init_right_pose;
    std::optional<std::array<double,6>> init_robot_pose;
    Eigen::Matrix4d T_offset_right = Eigen::Matrix4d::Identity();



    // -------------------------------------
    // BUTTON HANDLING HELPER 
    // -------------------------------------
    // handles rising edge detection for buttons 
    std::unordered_map<std::string, bool> prev_buttons;
    auto button_pressed = [&](const VRState& st, const std::string& name) {
        auto it = st.buttons.find(name);
        if (it == st.buttons.end()) return false;
        if (!std::holds_alternative<bool>(it->second)) return false;

        bool current = std::get<bool>(it->second);
        bool previous = prev_buttons[name];
        prev_buttons[name] = current;
        return (!previous && current);  // rising edge
    };



    // ---------------------------
    // Main teleop loop (~200 Hz)
    // ---------------------------
    while (true) {

        // Read a full VRState packet manually
        auto frame_opt = vr_manager.poll_manual();
        if (!frame_opt.has_value()) {
            continue;
        }

        const VRState& frame = *frame_opt;


        // -------------------------------------
        // BUTTON HANDLING
        // -------------------------------------
        // Track previous button states (bool-type buttons only)
        // Toggle pause/resume on A
        if (button_pressed(frame, "a")) {
            pause_teleop = !pause_teleop;
            std::cout << (pause_teleop ? "Teleop PAUSED\n" : "Teleop RESUMED\n");
        }


        // B returns robot to idle pose
        if (button_pressed(frame, "b")) {
            std::cout << "Returning to IDLE pose.\n";
            driver.set_arm_positions(IDLE_POSE, 3.0, true);
            break;
        }





        // -------------------------------------
        // GRIPPER TRIGGER
        // -------------------------------------
        {
            auto it = frame.buttons.find("right_trigger");
            if (it != frame.buttons.end() && std::holds_alternative<double>(it->second)) {
                double trig_val = std::get<double>(it->second);
                driver.set_gripper_position(trig_val * 0.04, 0.0, false);
            }
        }


        // -------------------------------------
        // RIGHT CONTROLLER (main teleop)
        // -------------------------------------
        if (!pause_teleop) {
            auto right_v6 = pose_to_vec6(frame.right_pose);
            if (right_v6.has_value()) {

                // controller robot alignment
                if (!init_right_pose.has_value()) {
                    init_robot_pose = driver.get_cartesian_positions();
                    init_right_pose = right_v6;

                    T_offset_right =
                        vec6_to_T(*init_robot_pose) *
                        vec6_to_T(*right_v6).inverse();

                    std::cout << "Teleop initialized.\n";
                }

                Eigen::Matrix4d Tq = vec6_to_T(*right_v6);
                Eigen::Matrix4d Tt = T_offset_right * Tq;
                auto robot_cmd = T_to_vec6(Tt);

                driver.set_cartesian_positions(
                    robot_cmd,
                    trossen_arm::InterpolationSpace::cartesian,
                    0.15, false
                );
            }
        }


        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }

    driver.set_arm_positions(IDLE_POSE, 3.0, true);
    return 0;
}

File: scripts/python/test_script2.py
import time
import math
import numpy as np

import pytrossen_vr
import trossen_arm

# ---------------------------
# Utils
# ---------------------------
def vec6_to_T(v6):
    p = np.array(v6[:3])
    rvec = np.array(v6[3:])
    angle = np.linalg.norm(rvec)
    R = np.eye(3)
    if angle > 1e-8:
        axis = rvec / angle
        R = rotation_matrix_from_axis_angle(axis, angle)
    T = np.eye(4)
    T[:3, :3] = R
    T[:3, 3] = p
    return T


def T_to_vec6(T):
    p = T[:3, 3]
    R = T[:3, :3]
    rvec = axis_angle_from_rotation_matrix(R)
    return [p[0], p[1], p[2], rvec[0], rvec[1], rvec[2]]


def pose_to_vec6(pose):
    if pose is None:
        return None
    return [
        pose.position[0],
        pose.position[1],
        pose.position[2],
        pose.rotation[0],
        pose.rotation[1],
        pose.rotation[2]
    ]


def rotation_matrix_from_axis_angle(axis, angle):
    # Rodrigues formula
    K = np.array([[0, -axis[2], axis[1]],
                  [axis[2], 0, -axis[0]],
                  [-axis[1], axis[0], 0]])
    return np.eye(3) + np.sin(angle) * K + (1 - np.cos(angle)) * (K @ K)


def axis_angle_from_rotation_matrix(R):
    angle = math.acos((np.trace(R) - 1) / 2)
    if abs(angle) < 1e-8:
        return np.zeros(3)
    rx = (R[2, 1] - R[1, 2]) / (2 * math.sin(angle))
    ry = (R[0, 2] - R[2, 0]) / (2 * math.sin(angle))
    rz = (R[1, 0] - R[0, 1]) / (2 * math.sin(angle))
    return np.array([rx, ry, rz]) * angle


# ---------------------------
# Robot setup
# ---------------------------
START_POSE = [0, math.pi/12, math.pi/12, 0, 0, 0]
IDLE_POSE  = [0, 0, 0, 0, 0, 0]

driver = trossen_arm.TrossenArmDriver()
driver.configure(trossen_arm.Model.wxai_v0,
                 trossen_arm.StandardEndEffector.wxai_v0_follower,
                 "192.168.1.2", True)
driver.set_arm_modes(trossen_arm.Mode.position)
driver.set_gripper_mode(trossen_arm.Mode.position)
driver.set_arm_positions(START_POSE, 3.0, True)
print("Robot ready.")


# ---------------------------
# VRManager setup (manual mode)
# ---------------------------
cfg = pytrossen_vr.VRManagerConfig()
cfg.server_port = 5432
vr_manager = pytrossen_vr.VRManager(cfg)
vr_manager.start()
print("VR Manager started in MANUAL mode.")


# ---------------------------
# Teleop state
# ---------------------------
pause_teleop = False
init_right_pose = None
init_robot_pose = None
T_offset_right = np.eye(4)

# Previous button states for rising edge detection
prev_buttons = {}

def button_pressed(frame, name):
    val = frame.buttons.get(name)
    if val is None or not isinstance(val, bool):
        return False
    current = val
    previous = prev_buttons.get(name, False)
    prev_buttons[name] = current
    return (not previous) and current


# ---------------------------
# Main loop (~200 Hz)
# ---------------------------
try:
    while True:
        frame = vr_manager.poll_manual()
        if frame is None:
            continue

        # ---------------------
        # BUTTON HANDLING
        # ---------------------
        if button_pressed(frame, "a"):
            pause_teleop = not pause_teleop
            print("Teleop PAUSED" if pause_teleop else "Teleop RESUMED")

        if button_pressed(frame, "b"):
            print("Returning to IDLE pose.")
            driver.set_arm_positions(IDLE_POSE, 3.0, True)
            break

        # ---------------------
        # GRIPPER TRIGGER
        # ---------------------
        trig_val = frame.buttons.get("right_trigger")
        if isinstance(trig_val, float):
            driver.set_gripper_position(trig_val * 0.04, 0.0, False)

        # ---------------------
        # RIGHT CONTROLLER (main teleop)
        # ---------------------
        if not pause_teleop:
            right_v6 = pose_to_vec6(frame.right_pose)
            if right_v6 is not None:
                if init_right_pose is None:
                    init_robot_pose = driver.get_cartesian_positions()
                    init_right_pose = right_v6
                    T_offset_right = vec6_to_T(init_robot_pose) @ np.linalg.inv(vec6_to_T(init_right_pose))
                    print("Teleop initialized.")

                Tq = vec6_to_T(right_v6)
                Tt = T_offset_right @ Tq
                robot_cmd = T_to_vec6(Tt)
                driver.set_cartesian_positions(robot_cmd,
                                               trossen_arm.InterpolationSpace.cartesian,
                                               0.15, False)

        time.sleep(0.005)

finally:
    driver.set_arm_positions(IDLE_POSE, 3.0, True)
